src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -1326,5 +1322,47 @@ public void ChainWithMaxReorgPlusExtraHeadersIsCalled_AnotherChainIsPresented_Co
                 }
             }
         }
+
+        /// <summary>
+        /// Issue 29 @ Peer presents at least two headers. Those headers will be connected to the tree. 
+        /// Then we save the first such connected block to variable X and simulate block downloaded for both blocks.
+        /// Then the peer is disconnected, which removes its chain from the tree. 
+        /// Partial validation succeeded is then called on X.
+        /// </summary>
+        [Fact]
+        public void ChainedHeaderIsRemovedFromTheTree_PartialValidationSucceededCalled_NothingIsReturned()
+        {
+            // Chain header tree setup. Initial chain has 5 headers.
+            // Example: h1=h2=h3=h4=h5.
+            const int initialChainSize = 5;
+            TestContext ctx = new TestContextBuilder()
+                .WithInitialChain(initialChainSize)
+                .Build();
+            ChainedHeaderTree cht = ctx.ChainedHeaderTree;
+            ChainedHeader chainTip = ctx.InitialChainTip;
+
+            // Extend chaintip with 2 headers, i.e. h1=h2=h3=h4=h5=h6=h7.
+            const int chainExtension = 2;
+            chainTip = ctx.ExtendAChain(chainExtension, chainTip);
+            
+            // Peer 1 presents these 2 headers.
+            const int peer1Id = 1;
+            List<BlockHeader> listOfHeaders = ctx.ChainedHeaderToList(chainTip, chainExtension);
+            ConnectNewHeadersResult connectionResult = cht.ConnectNewHeaders(peer1Id, listOfHeaders);
+            ChainedHeader consumedHeader = connectionResult.Consumed;
+
+            // Download both blocks.
+            ChainedHeader firstPresentedHeader = consumedHeader.Previous;
+            cht.BlockDataDownloaded(consumedHeader, chainTip.Previous.Block);
```

Comment:
consumedHeader here is 7, but you take block from chainTip.Previous which is 6. 
On the next line 
```
cht.BlockDataDownloaded(firstPresentedHeader, chainTip.Block);
```
firstPresentedHeader is 6, chainTip is 7.


It doesn't affect the test because min validation (that checks if block hash == chainedheader.Hash) is invoked on CM level, but still would be nice to make it logically correct.

Reply:
fixed